### PR TITLE
fix(config): unwrap quoted config values

### LINKED ISSUE #14427: pnpm 12 preserves quotes around interpolated .npmrc auth tokens
### Verify latest release

- [x] I verified that the issue exists in the latest pnpm release

### pnpm version

12.2.1

### Which area(s) of pnpm are affected? (leave empty if unsure)

CLI

### Link to the code that reproduces this issue or a replay of the bug

_No response_

### Reproduction steps

1. Create .npmrc:
```
@scope:registry=https://registry.example/
//registry.example/:_authToken="${TOKEN}"
```
2. Add a private @scope/* dependency.
3. Export a valid synthetic/test token:
```
export TOKEN=...
```
4. Run:
```
pnpm install
```

### Describe the Bug

After bumping pnpm from v11.13.0 to v12.2.1, installs of private packages from our internal registry fail with `ERR_PNPM_FETCH_401`

Both of the versions are using the same global `.npmrc` file.

```
//registry.example/:_authToken="${TOKEN}"
```

With `TOKEN=secret`

- pnpm 11 parses the value as `secret`
- pnpm 12 parses the values as `"secret"`

It seems the quotes are becoming part of the value and are causing auth to fail with `401 Unauthorized`.

To help repro I've used the unit tests beneath in both v11 and v12.

The TypeScript implementation passes this test at `pnpm11/config/reader/test/index.ts`:

```ts
test('user .npmrc expands a quoted auth token without retaining quotes', async () => {
  prepare()

  fs.mkdirSync('user-home')
  const userconfig = path.resolve('user-home', '.npmrc')
  fs.writeFileSync(
    userconfig,
    '//registry.example/:_authToken="${TOKEN}"\n',
    'utf8'
  )

  const { config } = await getConfig({
    cliOptions: { userconfig },
    env: { ...env, TOKEN: 'secret' },
    packageManager: {
      name: 'pnpm',
      version: '1.0.0',
    },
  })

  expect(config.authConfig['//registry.example/:_authToken']).toBe('secret')
})
```

The equivalent Rust implementation fails at `pnpm/crates/config/src/npmrc_auth/tests.rs`:

```rs
#[test]
fn from_ini_expands_quoted_auth_env_placeholder_without_quotes() {
    static_env!(Env, &[("TOKEN", "secret")]);

    let auth = NpmrcAuth::from_ini::<Env>(
        "//registry.example/:_authToken=\"${TOKEN}\"\n",
        Path::new(""),
    );

    assert_eq!(
        default_auth_token(&auth, "//registry.example/"),
        Some(Some("secret")),
    );
}
```

### Expected Behavior

Expected: pnpm 12.2.1 removes the quotes the same as pnpm 11 and authenticates successfully.
Actual: pnpm 12.2.1 is keeping the quotes as part of the token, causing `401 Unauthorized`

### Which Node.js version are you using?

24.14.1

### Which operating systems have you used?

- [x] macOS
- [ ] Windows
- [ ] Linux

### If your OS is a Linux based, which one it is? (Include the version if relevant)

_No response_

### PR DESCRIPTION
**Pull request by zkochan**
## Summary

- Decode matching single- and double-quoted `.npmrc` values before environment-variable expansion in the Rust CLI.
- Preserve the project-level auth secret guard after decoding, including when a double-quoted JSON escape produces the placeholder.
- Add regression and edge-case tests plus a patch changeset for `pacquet`.

The TypeScript CLI already strips matching quotes through `read-ini-file` and `ini`, so no TypeScript code change is needed.

Closes pnpm/pnpm#14427.

## Squash Commit Body

```text
Decode matching single- and double-quoted `.npmrc` values before environment-variable expansion in the Rust config parser. This matches the TypeScript parser while ensuring the project-level auth secret guard evaluates the decoded value, including JSON-escaped placeholders.

Closes pnpm/pnpm#14427.
```

## Checklist

- [x] I checked the referenced issue and verified that none of the PRs
  already linked to it solves it.
- [x] The change is implemented in both the TypeScript CLI and the Rust
  `pnpm/` port, or the description notes what still needs porting.
- [x] Added a changeset (`pnpm changeset`) if this PR changes any published
  package. Keep it short and written for pnpm users — it becomes a release note.
- [x] Added or updated tests.

---
Written by an agent (Codex, gpt-5.6-sol).

---
<a href="https://app.blacksmith.sh/pnpm/codesmith/pnpm/pr/14438"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with [code]smith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1790898344&installation_model_id=426870&pr_number=14438&repository=pnpm%2Fpnpm&return_to=https%3A%2F%2Fgithub.com%2Fpnpm%2Fpnpm%2Fpull%2F14438&signature=dd26c5276a7ea841cb0f4069efc371302e1dae1bd5bf96cf5c687c18a46c5c8c"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with [code]smith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith-bot</code> with what you need. Autofix is disabled.</sup>

## Summary by CodeRabbit

- **Bug Fixes**
  - Fixed authentication tokens in `.npmrc` files when values are surrounded by single or double quotes.
  - Environment-expanded tokens now correctly preserve supported quoting and escape sequences.
  - Quoted authentication settings in untrusted project configuration remain blocked and generate warnings.

- **Tests**
  - Added coverage for quoted tokens, environment substitution, escaped placeholders, and security protections.

### CODE REVIEW

File: pnpm/crates/config/src/npmrc_auth.rs [hunk 1]
```diff
@@ -866,6 +867,18 @@ impl NpmrcAuth {
     }
 }
 
+fn decode_ini_value(value: &str) -> Cow<'_, str> {
+    if value.starts_with('\'') && value.ends_with('\'') {
+        Cow::Borrowed(
+            value.strip_prefix('\'').and_then(|value| value.strip_suffix('\'')).unwrap_or(""),
+        )
+    } else if value.len() >= 2 && value.starts_with('"') && value.ends_with('"') {
+        serde_json::from_str::<String>(value).map_or(Cow::Borrowed(value), Cow::Owned)
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/config/src/npmrc_auth.rs:876]: Thanks for flagging this. No change is needed here: the TypeScript parser also retains the matching double quotes when JSON decoding fails. Reproducing this with the exact `ini` dependency installed under `pnpm11/config/commands`, `ini.parse('x="abc\\d"').x` evaluates to `"abc\\d"`, including both quote characters. Its `unsafe` helper catches the `JSON.parse` error and returns the original value.

The fallback in commit cb54a8d7577c5f7e0a9b3df4981f7f0a88f229f2 deliberately preserves that behavior, so the Rust and TypeScript CLIs remain aligned for invalid escapes.

---
Written by an agent (Codex, gpt-5.6-sol).

File: pnpm/crates/config/src/npmrc_auth.rs [hunk 2]
```diff
@@ -866,6 +867,18 @@ impl NpmrcAuth {
     }
 }
 
+fn decode_ini_value(value: &str) -> Cow<'_, str> {
+    if value.starts_with('\'') && value.ends_with('\'') {
```
[the remark this answers is not in the document]
Comment (zkochan) [reply pnpm/crates/config/src/npmrc_auth.rs:871]: The current behavior is intentional and matches the exact TypeScript `ini` parser. Reproducing the edge cases with the installed dependency gives:

```text
"'"  => ""
"''" => ""
"\""  => "\""
"\"\"" => ""
```

For a single-quoted value, `unsafe` removes the first and last character before attempting `JSON.parse`, so a lone single quote becomes empty. A lone double quote does not satisfy `isQuoted` and remains verbatim. The `("'", "")` case added in commit cb54a8d7577c5f7e0a9b3df4981f7f0a88f229f2 explicitly protects this TypeScript parity.

---
Written by an agent (Codex, gpt-5.6-sol).

**Comment by zkochan** [comment]
The two Qodo changeset findings are not applicable. The current root `AGENTS.md` section “Changesets for the Rust products” requires a changeset for a user-visible Rust product change, targeting `pacquet`, and explicitly states: “Do not add `"pnpm"` to a Rust-only changeset.”

This PR changes only the Rust CLI because the TypeScript CLI already has the correct behavior. The `pacquet`-only patch changeset in commit cb54a8d7577c5f7e0a9b3df4981f7f0a88f229f2 therefore follows the repository’s current product-specific instructions, so no changeset modification is needed.

---
Written by an agent (Codex, gpt-5.6-sol).
